frame/examples/offchain-worker-price-oracle/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+				// already did.
+				Err(MutateStorageError::ConcurrentModification(_)) => {
+					log::error!("OCW failed to acquire a lock.")
+				},
+			}
+		}
+	}
+
+	/// A public part of the pallet.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit new price to the list.
+		///
+		/// This method is a public function of the module and can be called from within
+		/// a transaction. It appends given `price` to current list of prices.
+		/// In our example the `offchain worker` will create, sign & submit a transaction that
+		/// calls this function passing the price.
+		///
+		/// This only works if the caller is in `Authorities`.
+		#[pallet::call_index(0)]
+		#[pallet::weight({0})]
+		pub fn submit_price(origin: OriginFor<T>, price: u32) -> DispatchResultWithPostInfo {
+			// Retrieve sender of the transaction.
+			let who = ensure_signed(origin)?;
+
+			match Self::is_authority(&who) {
+				true => Self::add_price(Some(who), price),
+				false => return Err(Error::<T>::NotAuthority.into()),
+			}
+
+			// Authorized OCWs don't need to pay fees
+			Ok(Pays::No.into())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight({0})]
+		pub fn add_authority(
+			origin: OriginFor<T>,
+			authority: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			ensure!(!Self::is_authority(&authority), Error::<T>::AlreadyAuthority);
+
+			let mut authorities = <Authorities<T>>::get();
+			match authorities.try_push(authority.clone()) {
+				Ok(()) => (),
+				Err(_) => return Err(Error::<T>::TooManyAuthorities.into()),
+			};
+
+			Authorities::<T>::set(authorities);
+
+			Ok(().into())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight({0})]
+		pub fn remove_authority(
+			origin: OriginFor<T>,
+			authority: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			ensure!(Self::is_authority(&authority), Error::<T>::NotAuthority);
+
+			let mut authorities = <Authorities<T>>::get();
+			match authorities.iter().position(|a| a == &authority) {
+				Some(index) => authorities.swap_remove(index),
+				None => return Err(Error::<T>::NotAuthority.into()),
+			};
+
+			Authorities::<T>::set(authorities);
+
+			Ok(().into())
+		}
+	}
+
+	/// Events for the pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Event generated when new price is accepted to contribute to the average.
+		NewPrice { price: u32, maybe_who: Option<T::AccountId> },
+	}
+
+	/// A vector of recently submitted prices.
+	///
+	/// This is used to calculate average price, should have bounded size.
+	#[pallet::storage]
+	#[pallet::getter(fn prices)]
+	pub(super) type Prices<T: Config> = StorageValue<_, BoundedVec<u32, T::MaxPrices>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub(super) type Authorities<T: Config> =
+		StorageValue<_, BoundedVec<T::AccountId, T::MaxAuthorities>, ValueQuery>;
+}
```

Comment:
I would move this code to the top of the pallet since it will be used by the dispatchable call logic